### ISSUE TITLE
Google Wins-Form Response Security

### ISSUE THREAD

**Issue by Abel-Zambrano** [hackforla/website#1980]
### Overview
As a developer, we need to protect our data. In the Google "Wins-form (Responses), we need to improve the security of sensitive data.

### Action Items
- [x] Research any strategies, tools, or solutions to protecting leaked API keys and secrets in our apps script.
- [x] As part of the research be sure to document either here or in our google drive:
  - [x] Things that have been tried
  - [x] Whether or not they are viable or a dead-end
  - [x] Make a list of requirements and limits to all viable solutions
- [x] In a meeting, discuss with the team to decide which solution is viable and can move to the demo stage.
- [x] If a solution is chosen, create a demo that fulfills these criteria:
  - [x] The keys or secrets are not visible at all, or readable by programmatic means without supplying some sort of secret token or passcode.
  - [x] The keys or secrets are readable by the apps script without exposing other keys or secrets.

### Resources/Instructions
To elaborate on the above items, we would want the apps script to behave somewhat like a GitHub secret token. A GH secret token is a special variable that GitHub supplies for every token. This token can be called, for example, in a GHA through a variable such as env.GITHUB_TOKEN or something similar. This makes it so that the exact token is normally not viewable (and attempts to console.log(env.GITHUB_TOKEN) will fail). So how is this token made in the first place? It is part of the settings of every repository, settings which only admins and owners can change and view. This means that the actual token itself is only known (theoretically) by someone with the required permission.

How does that relate to this task? First, we want our apps script to be able to read a token to perform it's tasks (ala GITHUB_TOKEN), but is not readable by anyone making edits to the apps script. Second, we want to be able to change or view the actual value of the token only if certain permissions are supplied (such as the password to an admin account).

**Comment by ExperimentsInHonesty** [comment]
@Abel-Zambrano  Is it possible to add more information to this issue?  https://github.com/hackforla/website/issues/1980  For instance, where is the folder that you are referencing.  what file contains the token, etc.  If the reason you are not putting the information in the issue is because you think its a security risk to put it in the issue.  Just put the content into the 1 password vault.  If you need help with that, please ask me in #admin channel

**Comment by Abel-Zambrano** [comment]
@ExperimentsInHonesty Hello, I added the details to the "1 Password Vault."

**Comment by Aveline-art** [comment]
> 1 Password Access for additional details.

@Abel-Zambrano Can you clarify what this means?

**Comment by Abel-Zambrano** [comment]
@Aveline-art I updated the issue. Let me know if more info is needed to clarify.

**Comment by Aveline-art** [comment]
* Progress: A new admin drive was created this weekend. The next step is to move the apps script to the drive.
* Blockers: None
* Availability: 1 hour
* ETA: Soon

**Comment by macho-catt** [comment]
### Currently Trying and Testing

### I. Store API key in a Google Doc
**Info:**

1. Need a secret folder in an admin google drive to store google doc (we have this already)
2. Need the ability to reference that google doc in Google App Script
   - If the wins google sheet and google app script are in the same google drive as the secret folder, we can use the DriveApp Class
   - If the wins google sheet and google app script are in a different drive, then we need to use a different class. One possible class is DocumentApp class and use openById to directly get the doc.
3. Preferably, encode the text in the doc and decode it in Google App Script to add an extra layer of protection


### II. Store API key Using Google Secret Manager
**Info:**

1. Need to enable Secret Manager API via Google Cloud Platform, then activate billing
2. Free tier allows six active secret versions
3. Google App Script currently does not have a library that provides direct access to Google Secret Manager (at least, I couldn't find any documentation on it nor did stackoverflow help). However, there is a workaround. When using the Secret Manager API, an OAuth Token needs to be provided. This can be obtained via the ScriptApp Class.
    - **Note:** To use the ScriptApp.getOAuthToken(), OAuth scopes need to be explicitly specified in the script's manifest file. However, doing so will require for all scopes for other Apps Scripts Services to be declared as well. Whenever a new scope is added, google will ask to re-review and re-authenticate the permissions for the script.


**Both are viable solutions.**

**Comment by macho-catt** [comment]
### Update:
We looked at both options, and although they can work, there are still some security concerns. For example, if a developer accidentally leaves a console.log or Logger.log on a variable that uses the key, then the key will still be leaked. We need a way to  hide the key from being viewed by console.log.

Some more ideas were floated, and this is a list of some of the ideas that could be helpful:

- Our current wins script is a container-bound script. That means that is currently attached to the wins-form (responses) sheet. The google apps script gets the same permission level as the sheet it is attached to. So if the permission level is set to "View Only," the google apps script is only set to "View Only" and cannot be edited nor can it be run
- We will generally set the permissions for devs working on google app script related issues to "View Only" and have them test and provide a demo on their local google drive first
- I saw an idea from this [article](https://community.airtable.com/t/api-key-protection-for-google-apps-script/37026). The idea is separating the library that calls the API from the wins script so that a developer cannot directly console.log and see the token
    - @Aveline-art and I also discussed something similar during our meeting

My current action item is to explore bullet point 3 and come up with a demo that uses ideas from it

Availability: 6 hours
ETA: 1 week

**Comment by macho-catt** [comment]
### Update:
I have a demo to present to the team tonight.

Currently, I have this following structure inspired by [this article](https://community.airtable.com/t/api-key-protection-for-google-apps-script/37026):

- A folder (let's call it hfla-test) in my own google drive to mimic the admin folder/drive
- A secret folder within hfla-test that will have strict permissions set. The encoded API key is in a google doc inside that folder
- A copy of the wins form and wins form response google sheet in hfla-test
- A container-bound google app script linked to the wins form response sheet. This google app script (let's call it testing-wins) has three functionalities:
    - A time-based trigger that updates test.json in my forked repo's testing-branch and creates a PR in my forked repo's gh-pages branch
    - A form submission trigger that opens a new issue about the new wins submission, and adds it to a project column
    - A form submission trigger that updates the 'Review' sheet of the wins form response google sheet (for admin/PMs)
    - ****Note:** I only revised/refactored the first two functionalities above
- A standalone google app script (let's call it gh-requests) in my google drive that acts as a library for GitHub API requests
    - At the moment, only I (the owner) can view and edit this script
    - The script contains the functions that retrieve and use the token for each API request
        - The three main functions are a get request, a post request, and a put request
    - testing-wins uses this library to make the API requests, but doesn't know or see what the API token is

**Comment by macho-catt** [comment]
### Update

- Fixed and moved functions in my prototype to be more secure. 
- Added descriptive info via console.log for each step so it can be debugged if something goes wrong (although this could definitely be expanded upon).

### To do

- Need to write documentation on everything for future devs / leads
- Need to implement the actual transfer to using the admin drive
    - Wins form responses and scripts need to move
    - API token needs to be moved
    - My prototype script needs to be implemented on the moved files
    - Test it all works

**Comment by macho-catt** [comment]
Did testing today. Everything works.

Final things to do:
 - get approval to move wins form and responses to admin drive
 - currently, the triggers for the google app script cannot be edited so to avoid having post requests for every daily trigger, I commented out those code so the script does not do anything. When we move over to the admin drive, I will set up everything under an admin account. I also reached out to Josh to delete those triggers for the time being.
 - finish documentation
